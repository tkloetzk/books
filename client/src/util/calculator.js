import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
