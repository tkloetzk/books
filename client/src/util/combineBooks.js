import forEach from 'lodash/forEach';
import mergeByKey from 'array-merge-by-key';
import isEmpty from 'lodash/isEmpty';

export function combineBooks(
  amazonBooks,
  googleBooks,
  goodreadsBooks,
  bookshelf
) {
  const combinedBooks = mergeByKey(
    'isbn',
    amazonBooks,
    googleBooks,
    goodreadsBooks
  );
  let duplicates = [];
  let duplicatedISBNs = [];
  forEach(combinedBooks, duplicatedBook => {
    return forEach(bookshelf, existingBook => {
      if (duplicatedBook.isbn === existingBook.isbn) {
        duplicatedBook.differences = this.compareDifferences(
          existingBook,
          duplicatedBook,
          []
        );
        // TODO: If duplicate but no differences exist, don't add but show notification
        if (duplicatedBook.differences.length) {
          duplicatedBook._id = existingBook._id;
          duplicates.push(duplicatedBook);
        } else {
          duplicatedISBNs.push({ isbn: duplicatedBook.isbn });
        }
      }
    });
  });
  return { combinedBooks, duplicates, duplicatedISBNs };
}
export function compareDifferences(oldBook, newBook, difference) {
  Object.keys(oldBook).forEach(key => {
    if (typeof oldBook[key] !== 'object') {
      if (
        oldBook[key] !== newBook[key] &&
        key !== '__v' &&
        key !== '_id' &&
        key !== 'adjustedRating' &&
        newBook[key] !== ''
      )
        difference.push({
          key,
          currentValue: oldBook[key],
          newValue: newBook[key],
        });
    } else {
      compareDifferences(oldBook[key], newBook[key], difference);
    }
  }, difference);

  return difference;
}

export default { combineBooks, compareDifferences };
