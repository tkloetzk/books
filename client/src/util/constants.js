export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
