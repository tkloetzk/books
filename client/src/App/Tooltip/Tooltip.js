import React from 'react';
import InitialIcon from '@material-ui/icons/RadioButtonUnchecked';
import DoneIcon from '@material-ui/icons/CheckCircle';
import ErrorIcon from '@material-ui/icons/Error';
import { LOADING_STATUSES } from '../../util/constants';
import CircularProgress from '@material-ui/core/CircularProgress';
import Typography from '@material-ui/core/Typography';

function TooltipProgress(props) {
  const { progress } = props;
  const green = { color: 'green' };
  if (progress === LOADING_STATUSES.initial)
    return <InitialIcon style={green} fontSize="small" />;

  if (progress === LOADING_STATUSES.loading) {
    return (
      <CircularProgress
        size={18}
        style={{ color: 'yellow', marginTop: '1px' }}
        thickness={4}
      />
    );
  }
  if (progress === LOADING_STATUSES.success) return <DoneIcon style={green} />;
  if (progress === LOADING_STATUSES.errored)
    return <ErrorIcon style={{ color: 'red' }} />;
}

export default function Tooltip({ content }) {
  if (!content) {
    return null;
  }
  return (
    <div style={{ width: 100 }}>
      {content.map(tooltip => (
        <span
          style={{
            display: 'flex',
            flexDirection: 'row',
            justifyContent: 'space-between',
          }}
          key={tooltip.label}
        >
          <Typography style={{ color: 'white' }}>{tooltip.label}</Typography>
          <TooltipProgress progress={tooltip.loading} />
        </span>
      ))}
    </div>
  );
}
