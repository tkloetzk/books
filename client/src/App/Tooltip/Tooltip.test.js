import React from 'react';
import Tooltip from './Tooltip';
import { shallow } from 'enzyme';
import forEach from 'lodash/forEach';
import { LOADING_STATUSES } from '../../util/constants';

describe('Tooltip', () => {
  let props;
  let wrapper;
  let instance;

  beforeEach(() => {
    props = { conent: { label: 'Amazon', loading: 'AMAZON_LOADING' } };
    wrapper = shallow(<Tooltip {...props} />);
    instance = wrapper.instance();
  });
  it('renders with no props', () => {
    wrapper = shallow(<Tooltip />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly with no content in props', () => {
    props = Object.assign({}, props, { content: '' });
    wrapper = shallow(<Tooltip {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
  it('renders with correct props', () => {
    forEach(LOADING_STATUSES, status => {
      props = Object.assign({}, props, {
        content: [{ label: 'Amazon', loading: status }],
      });
      wrapper = shallow(<Tooltip {...props} />);

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
