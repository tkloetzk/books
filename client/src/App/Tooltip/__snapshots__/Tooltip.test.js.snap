// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders correctly with no content in props 1`] = `""`;

exports[`Tooltip renders with correct props 1`] = `"<div style=\\"width:100px\\"><span style=\\"display:flex;flex-direction:row;justify-content:space-between\\"><p class=\\"MuiTypography-root-1 MuiTypography-body1-10\\" style=\\"color:white\\">Amazon</p><svg class=\\"MuiSvgIcon-root-37 MuiSvgIcon-fontSizeSmall-44\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"color:green\\"><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"></path><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path></svg></span></div>"`;

exports[`Tooltip renders with correct props 2`] = `"<div style=\\"width:100px\\"><span style=\\"display:flex;flex-direction:row;justify-content:space-between\\"><p class=\\"MuiTypography-root-1 MuiTypography-body1-10\\" style=\\"color:white\\">Amazon</p><svg class=\\"MuiSvgIcon-root-37\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"color:#64b5f6\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z\\"></path></svg></span></div>"`;

exports[`Tooltip renders with correct props 3`] = `"<div style=\\"width:100px\\"><span style=\\"display:flex;flex-direction:row;justify-content:space-between\\"><p class=\\"MuiTypography-root-1 MuiTypography-body1-10\\" style=\\"color:white\\">Amazon</p><div class=\\"MuiCircularProgress-root-46 MuiCircularProgress-colorPrimary-49 MuiCircularProgress-indeterminate-48\\" style=\\"width:18px;height:18px;color:yellow;margin-top:1px\\" role=\\"progressbar\\"><svg class=\\"MuiCircularProgress-svg-51\\" viewBox=\\"22 22 44 44\\"><circle class=\\"MuiCircularProgress-circle-52 MuiCircularProgress-circleIndeterminate-54\\" cx=\\"44\\" cy=\\"44\\" r=\\"20\\" fill=\\"none\\" stroke-width=\\"4\\"></circle></svg></div></span></div>"`;

exports[`Tooltip renders with correct props 4`] = `"<div style=\\"width:100px\\"><span style=\\"display:flex;flex-direction:row;justify-content:space-between\\"><p class=\\"MuiTypography-root-1 MuiTypography-body1-10\\" style=\\"color:white\\">Amazon</p><svg class=\\"MuiSvgIcon-root-37\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"color:green\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z\\"></path></svg></span></div>"`;

exports[`Tooltip renders with correct props 5`] = `"<div style=\\"width:100px\\"><span style=\\"display:flex;flex-direction:row;justify-content:space-between\\"><p class=\\"MuiTypography-root-1 MuiTypography-body1-10\\" style=\\"color:white\\">Amazon</p><svg class=\\"MuiSvgIcon-root-37\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\" style=\\"color:red\\"><path fill=\\"none\\" d=\\"M0 0h24v24H0z\\"></path><path d=\\"M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z\\"></path></svg></span></div>"`;

exports[`Tooltip renders with no props 1`] = `""`;
